File: data-structures/singly-linked-list/singly-linked-list.js


class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class SinglyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    Node(value) {
        return {
            value,
            next: null,
        }
    }

    push(value) {
        const newNode = this.Node(value);
        // condition checks if list is empty
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length ++;
        return this;
    }

    pop() {
        if(!this.head) return undefined;
        if(this.head === this.tail) {
            this.head = null;
            this.tail = null;
            list.length = 0;
            return undefined;
        }

        const oldTail = this.tail;
        let node = this.head;

        while(node.next !== this.tail) {
            node = node.next;
        }

        this.tail = node;
        this.tail.next = null;
        this.length --;
        return oldTail;
    }

    shift() {
        const oldHead = this.head;
        if(!this.head) return undefined;
        if(this.head === this.tail) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
        }

        this.length --;
        return oldHead;
    }

    unshift(value) {
        const newNode = this.Node(value);
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }

        this.length ++;
        return this;
    }

    get(index) {
        if(index < 0 || index >= this.length) return undefined;
        let node = this.head;

        for(let i = 0; i < index; i++) {
            node = node.next;
        }

        return node;
    }

    set(index, value) {
        let node = this.get(index);
        if(!node) return undefined;

        node.value = value;
        return this;
    }

    insert(index, value) {
        if(index < 0 || index > this.length) return false;
        if(index === 0) {
            this.unshift(value);
            return true;
        } else if(index === this.length) {
            this.push(value);
            return true;
        }

        let newNode = this.Node(value);
        let prevNode = this.get(index - 1);
        let nextNode = prevNode.next;

        newNode.next = nextNode;
        prevNode.next = newNode;
        this.length ++;
        return true;
    }

    remove(index) {
        if(index < 0 || index > this.length) return false;
        if(index === 0) {
            this.shift();
            return true;
        } else if(index === this.length) {
            this.pop();
            return true;
        }
        
        let prevNode = this.get(index - 1);
        prevNode.next = prevNode.next.next;
        this.length --;
        return true;
    }

    reverse() {
        if(!this.head) return this;

        for(let i = 0; i < this.length ; i++) {
            this.insert(i, this.pop().value);
        }

        return this;
    }
}


const list  = new SinglyLinkedList();

list.push(2);
list.pop();
list.push(3);
list.push(1);
list.shift();
list.shift();
list.unshift("!");
list.unshift("bro");
list.unshift("story");
list.unshift("cool");
list.set(3, ".");
list.insert(3, ".....");
list.remove(0);
list.reverse()
console.log(JSON.stringify(list, null, 2));

File: data-structures/singly-linked-list/singly-linked-list.py


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def Node(self, value):
        class NODE:
            def __init__(self, value):
                self.value = value;
                self.next = None;

        return NODE(value)
    

    def push(self, value):
        new_node = self.Node(value)
        # case if list is empty
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

        self.length = self.length + 1
        return self
        

    def pop(self):
        old_tail = self.tail
        if not self.head:
            return None
        elif self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            node = self.head
            while node.next != self.tail:
                node = node.next
            
            self.tail = node
            self.tail.next = None

        self.length = self.length - 1
        return old_tail

    def unshift(self, value):
        new_node = self.Node(value)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        
        self.length = self.length + 1
        return self;

    def shift(self):
        old_head = self.head
        if not self.head:
            return None
        elif self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
        
        self.length = self.length - 1
        return old_head



new_list = SinglyLinkedList()

new_list.push("h")
new_list.push("ha")
new_list.push("hah")
new_list.push("haha")
new_list.pop()
new_list.pop()
new_list.unshift("a")
new_list.shift()

print(new_list.head.next.value)
